octoprint_firmwareupdate/__init__.py

```diff
@@ -64,110 +72,112 @@ def on_after_startup(self):
         self._start_update(True)
 
     def _start_update(self, onstartup=False):
+        # Make sure printer is disconnected before continuing
+        self._printer.disconnect()
+
         self._update_firmware_init_thread = Thread(target=self._update_firmware_init, args=(onstartup,))
         self._update_firmware_init_thread.daemon = True
         self._update_firmware_init_thread.start()
 
     def checkStatus(self):
         while True:
-            update_result = open(os.path.join(os.path.expanduser('~'), 'Marlin/.build_log')).read()
-            if 'No device matching following was found' in update_result:
-                self._logger.info("Failed update...")
-                self._update_status(False, "error", "A connected device was not found.")
-                self._clean_up()
-                break
-            elif 'FAILED' in update_result:
-                self._logger.info("Failed update...")
-                self._update_status(False, "error")
-                self._clean_up()
-                break
-            elif 'bytes of flash verified' in update_result and 'avrdude done' in update_result:
-                self._logger.info("Successful update!")
-                for line in update_result.splitlines():
-                    if "Reading" in line:
-                        self.read_time = self.find_between(line, " ", "s")
-                        self.completion_time += float(self.read_time)
-                    elif "Writing" in line:
-                        self.write_time = self.find_between(line, " ", "s")
-                        self.completion_time += float(self.write_time)
-
-                self._update_status(False, "completed", round(self.completion_time, 2))
-                self._clean_up()
-                break
-            elif 'ReceiveMessage(): timeout' in update_result:
-                self._logger.info("Update timed out. Check if port is already in use!")
-                self._update_status(False, "error", "Device timed out. Please check that the port is not in use!")
-
-                p = psutil.Process(self.updatePID)
-                for child in p.children(recursive=True):
-                    child.kill()
-                    p.kill()
-                self._clean_up()
-                break
-            sleep(1)
+            with open(os.path.expanduser('~/Marlin/.build_log')) as f:
+                update_result = f.read()
+                if 'No device matching following was found' in update_result:
```

Comment:
I don't think the following `if`s and `elif`s need to be inside the context handler's block.